### PR TITLE
ch9344: init at 1.9

### PR DESCRIPTION
**Pull request by MakiseKurisu**
###### Description of changes

This package provides kernel module for WinChipHead CH9344/CH348 USB To Multi Serial Ports controller.

The driver source code, along with the chipset datasheets, can be found [here](https://www.wch.cn/downloads/CH9344SER_LINUX_ZIP.html) (language warning: Chinese).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I'm new to NixOS so I don't know how to add out-of-tree modules from PR/local git repo to `configuration.nix`. I'm able to include the package to `environment.systemPackages`:

```
...
  nixpkgs.config = {
    # Allow unfree packages
    allowUnfree = true;
    packageOverrides = pkgs: {
      wip = import (fetchGit { url = "/home/user/nixpkgs"; shallow = true;}) { config = config.nixpkg>
    };
  };
...
  environment = {
    systemPackages = with pkgs; [
        wip.linuxKernel.packages.linux_5_15.ch9344
    ...
    ];
  };
```
<s>However, I don't know how to add `ch9344` to `boot.extraModulePackages`.</s> Figured it out and [Wiki updated](https://nixos.wiki/index.php?title=Linux_kernel#Loading_out-of-tree_kernel_modules).

ko generated according to [`Developing out-of-tree kernel modules`](https://nixos.wiki/wiki/Linux_kernel#Developing_out-of-tree_kernel_modules) is loadable and working with real hardware (I have a CH348 board). I also tested [patchPhase](https://nixos.wiki/wiki/Nixpkgs/Create_and_debug_packages#Using_nix-shell_for_package_development) to make sure patch is applied when building against 6.1 kernel.

### CODE REVIEW

File: pkgs/os-specific/linux/ch9344/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    hash = "sha256-g55ftAfjKKlUFzGhI1a/O7Eqbz6rkGf1vWuEJjBZxBE=";
+  };
+
+  patches = lib.optionals (lib.versionAtLeast kernel.modDirVersion "6.1") [
+    # https://github.com/torvalds/linux/commit/a8c11c1520347be74b02312d10ef686b01b525f1
+    ./fix-incompatible-pointer-types.patch
+  ];
+
+  sourceRoot = "${src.name}/driver";
+  hardeningDisable = [ "pic" ];
```
Comment (GabrielDougherty) [inline pkgs/os-specific/linux/ch9344/default.nix:19]: Would be good to leave a comment saying why this was needed

**Review by GabrielDougherty — APPROVED** [review]

File: pkgs/os-specific/linux/ch9344/default.nix (hunk 1 above)
Comment (RadxaYuntian) [reply pkgs/os-specific/linux/ch9344/default.nix:19]: I took that line from [here](https://nixos.wiki/wiki/Linux_kernel#Developing_out-of-tree_kernel_modules), which says:
```
For kernel modules it is necessary to disable pic in compiler hardenings as the kernel need different compiler flags. 
```

So it sounds to me that this is required and expected when creating a package for out-of-tree modules. For reference I use `r8125` package as the starting point and that package contains the same line without any comment. Same for a few other kernel modules (exfat, rtl8*) that I just checked right now.

Please let me know if you still think some explanation is needed here.

Comment (MakiseKurisu) [reply pkgs/os-specific/linux/ch9344/default.nix:19]: I took that line from [here](https://nixos.wiki/wiki/Linux_kernel#Developing_out-of-tree_kernel_modules), which says:
```
For kernel modules it is necessary to disable pic in compiler hardenings as the kernel need different compiler flags. 
```

So it sounds to me that this is required and expected when creating a package for out-of-tree modules. For reference I use `r8125` package as the starting point and that package contains the same line without any comment. Same for a few other kernel modules (exfat, rtl8*) that I just checked right now.

Please let me know if you still think some explanation is needed here.

Comment (GabrielDougherty) [reply pkgs/os-specific/linux/ch9344/default.nix:19]: Ok, it's fine

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -9156,6 +9156,11 @@
     githubId = 115218;
     name = "Felix Richter";
   };
+  MakiseKurisu = {
+    github = "MakiseKurisu";
```
[the remark this answers is not in the document]
Comment (RadxaYuntian) [reply maintainers/maintainer-list.nix:9160]: `githubId` is provided below. Isn't that [enough](https://github.com/NixOS/nixpkgs/blob/cefb9fda9fec9b56c4d51f8c368dfea96c851eba/maintainers/maintainer-list.nix#L7)?

[the remark this answers is not in the document]
Comment (MakiseKurisu) [reply maintainers/maintainer-list.nix:9160]: `githubId` is provided below. Isn't that [enough](https://github.com/NixOS/nixpkgs/blob/cefb9fda9fec9b56c4d51f8c368dfea96c851eba/maintainers/maintainer-list.nix#L7)?

File: pkgs/os-specific/linux/ch9344/default.nix [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  nativeBuildInputs = kernel.moduleBuildDependencies;
+
+  KERNELDIR="${kernel.dev}/lib/modules/${kernel.modDirVersion}/build";
+
+  buildPhase = "make -C $KERNELDIR M=$(pwd) modules";
```
[the remark this answers is not in the document]
Comment (MakiseKurisu) [reply pkgs/os-specific/linux/ch9344/default.nix:24]: I recalled some build issues with standard `buildPhase`. I'll check this again.

[the remark this answers is not in the document]
Comment (MakiseKurisu) [reply pkgs/os-specific/linux/ch9344/default.nix]: Removed `buildPhase` using the same techinque as in [`r8125`](https://github.com/NixOS/nixpkgs/blob/master/pkgs/os-specific/linux/r8125/default.nix#L23-L30).

File: maintainers/maintainer-list.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply maintainers/maintainer-list.nix:9160]: Well, then we have only github to reach you and if that isn't working we can't but sure.

**Pull request merged**